File: Python/112_Path_Sum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def hasPathSum(self, root: TreeNode, sum: int) -> bool:
        if root == None:
            return False
        if root.left == None and root.right == None:
            return True if sum - root.val == 0 else False
        else:
            a = False
            b = False
            if root.left != None:
                a = self.hasPathSum(root.left, sum-root.val)
            if root.right != None:
                b = self.hasPathSum(root.right, sum-root.val)
        return a or b

File: C++/112_Path_Sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool hasPathSum(TreeNode* root, int sum) {
        if(root == NULL) return false;
        bool a = false;
        bool b = false;
        if(root->left == NULL && root->right == NULL){
            return sum - root->val == 0 ? true : 0;
        } else {
            if(root->left != NULL){
                a = hasPathSum(root->left, sum-root->val);
            }
            if(a == false && root->right != NULL){
                b = hasPathSum(root->right, sum-root->val);
            }
        }
        return a || b;
    }
};